packages/orm/src/collections/CustomReference.ts

```diff
@@ -0,0 +1,126 @@
+import { Entity, IdOf, LoadHint, Loaded } from "../EntityManager";
+import { AbstractRelationImpl } from "./AbstractRelationImpl";
+import { ensureNotDeleted, fail, Reference } from "../index";
+
+export type CustomReferenceOpts<
+  T extends Entity,
+  U extends Entity,
+  H extends LoadHint<T>,
+  N extends never | undefined
+> = {
+  load: (entity: T) => Promise<void>;
+  get: (entity: Loaded<T, H>) => U | N;
+  set?: (entity: Loaded<T, H>, other: U) => void;
+  setFromOpts?: (entity: Loaded<T, H>, other: U) => void;
+  isSet?: (entity: Loaded<T, H>) => boolean;
+};
+
+export class CustomReference<T extends Entity, U extends Entity, H extends LoadHint<T>, N extends never | undefined>
+  extends AbstractRelationImpl<U>
+  implements Reference<T, U, N> {
+  private loadPromise: Promise<void> | undefined;
+  private _isLoaded = false;
+  constructor(public entity: T, public fieldName: keyof T, private opts: CustomReferenceOpts<T, U, H, N>) {
+    super();
+  }
+
+  private filterDeleted(entity: U | N, opts?: { withDeleted?: boolean }): U | N {
+    return opts?.withDeleted === true || entity === undefined || !entity.isDeletedEntity ? entity : (undefined as N);
+  }
+
+  private doGet(opts?: { withDeleted?: boolean }): U | N {
+    ensureNotDeleted(this.entity, { ignore: "pending" });
```

Comment:
Separately, do all of our `ensureNotDeleted` pass ignore pending at this point? If they're not, should they/should the param go away and it just always ignores pending?

Reply:
Some of our calls do not pass `ignore: "pending"`.  For example, o2m's `add` and `set` both still just have `ensureNotDeleted(this.entity)`, which is, imo, correct.